Flow/flow.h

```diff
@@ -0,0 +1,415 @@
+  virtual ~SubflowRunner() = default;
+};
+template <>
+struct SubflowRunner<InputPolicy::DoesNotAccept, OutputPolicy::Emits> : CanBeSpawnedToEmit {
+  virtual ~SubflowRunner() = default;
+};
+template <>
+struct SubflowRunner<InputPolicy::Accepts, OutputPolicy::DoesNotEmit> : CanBeSpawnedToAccept {
+  virtual ~SubflowRunner() = default;
+};
+template <>
+struct SubflowRunner<InputPolicy::Accepts, OutputPolicy::Emits> : CanBeSpawnedToAccept, CanBeSpawnedToEmit {
+  virtual ~SubflowRunner() = default;
+};
+
+// General fields for each subflow, complete or incomplete.
+template <InputPolicy INPUT, OutputPolicy OUTPUT>
+class Subflow : public SubflowRunner<INPUT, OUTPUT> {
+ public:
+  explicit Subflow(const std::string& description) : description_(description) {}
+
+  // No subflow should be left hanging.
+  virtual ~Subflow() {
+    if (still_hanging && !describe_called) {
+      // Too bad we can't throw an exception here. -- D.K.
+      // TODO(dkorolev): Unit-test this with a singleton policy with ostream and std::exit().
+      std::cerr << "Current leaked:\n" << Describe() << "\nCall `.Flow()`, etc., don't leave it hanging.\n";
+      std::exit(-1);
+    }
+  }
+
+  // The external, user-facing `Describe()` describes this subflow in human-readable text.
+  // Applies beautification, such as "... | Logic()" and/or "Logic() | ...", for to-be-added inputs/outputs.
+  // Does not not mark this subflow as used, but marks it as the one that was described.
+  std::string Describe() const {
+    MarkAsDescribeCalled();
+    return BeautifyDescription<INPUT, OUTPUT>::run(description_);
+  }
+
+  // The internal method to describe the subflow. Used to construct textual description of the complex ones.
+  // Marks the subflow as not hanging, as it is now part of a larger one.
+  std::string DescribeWithoutBangs() const {
+    MarkAsNotHanging();
+    return description_;
+  }
+
+  // Marks this subflow as "used". Surpasses error message on destruction. Should happen at most once.
+  void MarkAsNotHanging() const {
+    assert(still_hanging);
+    still_hanging = false;
+  }
+
+  // Marks this subflow as "described". Surpasses error message on destruction.
+  // Can happen multiple times, but not after the flow was run.
+  void MarkAsDescribeCalled() const {
+    assert(still_hanging);
+    describe_called = true;
+  }
+
+ private:
+  const std::string description_;
+  mutable bool still_hanging = true;
+  mutable bool describe_called = false;
+};
+
+typedef Subflow<InputPolicy::DoesNotAccept, OutputPolicy::DoesNotEmit> SubflowE2E;
+typedef Subflow<InputPolicy::DoesNotAccept, OutputPolicy::Emits> SubflowLHS;
+typedef Subflow<InputPolicy::Accepts, OutputPolicy::DoesNotEmit> SubflowRHS;
+typedef Subflow<InputPolicy::Accepts, OutputPolicy::Emits> SubflowVIA;
+
+// Base classes for user code definition.
+class SetEmitHelper;
+class EmitterImpl {
+ protected:
+  // `emit(message)`: Called by user code.
```

Comment:
Would `// `emit()` is the implementation of `emit(...)` called by the user from their code.` read better?

It will accept a templated type soon, `int` is just for testing.